ethcore/src/state/account.rs

```diff
@@ -59,6 +59,10 @@ pub struct Account {
 	// LRU Cache of the trie-backed storage.
 	// This is limited to `STORAGE_CACHE_ITEMS` recent queries
 	storage_cache: RefCell<LruCache<H256, H256>>,
+	// LRU Cache of the trie-backed storage for original value.
+	// This is only used when the initial storage root is different compared to
+	// what is in the database. That is, it is only used for new contracts.
+	original_storage_cache: Option<(H256, RefCell<LruCache<H256, H256>>)>,
```

Comment:
Yeah, `original_storage_cache` is a lot like `storage_root` + `storage_cache`. However, the usage for them are really different so I'm not sure if refactoring would save code.

And TBH, after we reviewed and merged this PR, I think the next reasonable step would be to speed up finishing #9427 -- the current checkpoint logic is not easy to follow, so moving the code would only help a little. If we make it so that account/storage changes struct are cheap to clone (through PDS), then I think checkpoint logic can be simplified a lot.